depcheck/depcheck.go

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package depcheck defines a test utility for ensuring certain packages don't
+// take on heavy dependencies.
+package depcheck
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"testing"
+
+	"golang.org/x/tools/go/packages"
+)
+
+type node struct {
+	importpath string
+	consumers  map[string]struct{}
+}
+
+type graph map[string]node
+
+func (g graph) Contains(name string) bool {
+	_, ok := g[name]
+	return ok
+}
+
+func (g graph) Path(name string) []string {
+	n := g[name]
+	// Base case.
+	if len(n.consumers) == 0 {
+		return []string{name}
+	}
+	// Inductive step.
+	consumers := make(sort.StringSlice, 0, len(n.consumers))
+	for k := range n.consumers {
+		consumers = append(consumers, k)
+	}
+	consumers.Sort()
+	base := g.Path(consumers[0])
+	if len(base) > 1 { // Don't decorate the first entry, which is always an entrypoint.
+		if len(consumers) > 1 {
+			// Attach other consumers to the last entry in base.
```

Comment:
can you show in comment an example of a string you're building?

Reply:
done